### ISSUE TITLE
Effects for input devices / streams from microphone

### ISSUE THREAD

**Issue by Bleuzen** [wwmm/pulseeffects#28]
Hi, I have a suggestion.
I would like to use this effects for my input (mic).
During playing around with PulseEffects and PulseAudio / pavucontrol ... I got it already working, that I can hear myself and apply the effects.
But I don't know how I can not hear myself and rout this into other apps, which want to record my voice ... and it would be great if this would be a feature instead of a random 'playing around' thing..

**Comment by wwmm** [comment]
Hi, what kind of applications would be recording? 

In any case it should be possible to implement this but it will require some work. The current PulseEffects code really has a bias towards processing applications output. I will think about the best way to switch between modes.

**Comment by Bleuzen** [comment]
The kind of applications which would be recording would be mainly voip, for example: Discord, TeamSpeak, Skype (and OBS (for streaming))

It also would be nice if we not only can switch between modes, but use PulseEffects for input and output at the same time.
Or, if you implement a switch, maybe multiple instances can do this.. But thats up to you how you do this.

Good luck and this would be a realy nice feature ;)

**Comment by wwmm** [comment]
The problem with multiple instances is that since the beginning  I wanted PulseEffects to remember every change made to the sliders so I wouldn't have to load a preset every time I launched it. This was implemented saving all the values to a dconf key. If I allow multiple instances all of them will try to write values to the same dconf key. It will be a mess.

I don't know if applying effects for input and output at the same time will ever be possible because in this case people would probably want to apply different settings for input and output.

**Comment by Bleuzen** [comment]
Ok, I can live with that. I can switch to output effects when I´m listening to music and switch to input effects when I´m talking ... thats ok.

For the interface I think about maybe a tabbed one. One tab for "Output" (default tab when opening PulseEffects) and one "Input" tab.
If you switch to the Input tab, PulseEffects should disable all output effects and enable all input effects. Same (reversed) for switching to the Output tab back.

For the dconf key if people want different settings for In-/Output:
What about creating
/com/github/wwmm/pulseeffects/output
and
/com/github/wwmm/pulseeffects/input
?

**Comment by capricorn86** [comment]
+1 for this feature

**Comment by Xeed84** [comment]
+1 for voice modulation on Skype and other apps!

**Comment by wwmm** [comment]
Hi

How would you know the effects are good without listening to your own voice? Skype and other voip applications give you an option to listen to your own voice so you know it is good? Or it will be necessary to have an option in PulseEffects to listen to your own voice? (I do not use voip apps so I do not have much idea about the typical use cases)

**Comment by Bleuzen** [comment]
Most apps don't give you an option to listen to your own voice. But in my opinion, this is not needed. I hate to listen to my own voice with a little delay. It's hard to talk then as usual.
I prefer to use Audacity to make a short test recording and than listen to it. So I wouldn't need or use an option, to listen to myself.

But as I said, just my opinion. If it is easy for you to implement.. Why not? I haven't to use it.

**Comment by capricorn86** [comment]
I would like to use the functionality for live voice manipulation, mostly when using Skype. Another way to use it is when doing live streams. The optimal way for me would be if it was possible to save presets of configurations. I would then be able to use recordings (with Audacity for example) to listen to my voice and create my voice presets.

There are several applications to Windows, but I have not been able to find a good software for this to Ubuntu. Based on my research it looks like there are many people that would like a software with this feature. I believe that PulseEffects would be better than the Windows applications available and more than I would hope for, if it provides the same functionality currently supported in output.

**Comment by wwmm** [comment]
Hi,

I have just released PulseEffects 2.0. It is capable of applying effects for microphone. This is done at the same time effects are being applied to applications output (2 gstreamer pipelines running in parallel). In order to alternate between controls for microphone effects and general applications effects use the buttons beside the autovolume switch. 

This version is not compatible with presets saved with version 1.x because now presets for microphone effects are also saved. Sorry for that but there was no other way...

As I told before This is not a feature I use so feedback is not only welcome but also necessary.

**Comment by Bleuzen** [comment]
Hi, I tested it now with some apps and it works so far.
If I find something, I will create a new issues.
I think we can close this one.
Thanks :D

**Comment by capricorn86** [comment]
Awesome, I will try it now, Thanks!

**Comment by capricorn86** [comment]
@wwmm it works great. Thank you! 

Is it possible to add a pitch effect as well? It would be cool. I can create a new issue to track it if it is possible.

**Comment by wwmm** [comment]
As long as there is a gstreamer or ladspa plugin that does that it should be possible
